File: simulation-api/hsr/characters/asta.ts
/**
 * @character Asta
 * @role Support / Harmony (Fire)
 * @core_mechanics
 *   Charging stacks (max 5): +14% ATK per stack to ALL allies.
 *   Gained during basic/skill hits (1 per unique enemy + 1 if Fire Weakness).
 *   From 2nd turn onward, -3 stacks per turn start (-2 at E6).
 *   Ultimate: +50 SPD to all allies for 2 turns (tracked per ally via onAllyAction).
 *
 * Implemented mechanics:
 *   Talent    ECharging stacks ↁEall allies +14% ATK/stack (dynamic, updates on gain/loss)
 *   A2        EBasic ATK 80% chance to Burn for 3 turns (DoT = 50% of basic hit DMG)
 *   A4        EAll Fire-element allies +18% Fire DMG at battle start
 *   A6        EAsta DEF +6%/Charging stack (minor, not tracked in combat buffs)
 *   E1        ESkill fires 1 extra bounce (5 ↁE6 total hits)
 *   E2        EAfter Ult, skip next Charging stack reduction
 *   E3        ESkill +2, Talent +2 (via eidolon_level_boosts)
 *   E4        E+15% ERR when 2+ Charging stacks (~+3 on basic, ~+1 on skill)
 *   E5        EUltimate +2, Basic +1 (via eidolon_level_boosts)
 *   E6        ECharging stack reduction per turn: -3 ↁE-2
 *
 * Energy: Basic +20 (auto), Skill +6 (corrected from simulator +30 in onBeforeAction), Ult +5 (onUlt)
 * SPD buff expiry: decremented via onAllyAction (other allies) and onAfterAction (Asta herself).
 */

import type { CharacterKit, SimState, TeamMember, SimEnemy } from "../types.js";
import { calculateHsrDamage } from "../formulas.js";

const ATK_ID    = 'c987f652-6a0b-487f-9e4b-af2c9b51c6aa';
const CHAR_SPD_ID = '3e4b082d-7943-440d-ae2c-8d31b0a370be';

const ASTA_ID     = 'a8b9c0d1-e2f3-4567-89ab-cdef01234567';

// State keys
const CHARGING_KEY  = 'asta_charging';
const E2_SKIP_KEY   = 'asta_e2_skip_reduction';
const TURN_KEY      = 'asta_turn_count';
const MAX_CHARGING  = 5;

// ── Helpers ───────────────────────────────────────────────────────────────────

function getAsta(state: SimState): TeamMember | undefined {
  return state.team.find(m => m.characterId === ASTA_ID);
}

/**
 * Update Charging stacks and propagate the ATK buff delta to ALL team members.
 * Handles capping at MAX_CHARGING and clamping at 0.
 */
function setCharging(state: SimState, newStacks: number): void {
  const old = state.stacks[CHARGING_KEY] || 0;
  const capped = Math.min(Math.max(0, newStacks), MAX_CHARGING);
  const diff = (capped - old) * 14; // 14% ATK per stack
  if (diff !== 0) {
    state.team.forEach(m => {
      if (!m.isDowned) m.buffs.atk_percent += diff;
    });
  }
  state.stacks[CHARGING_KEY] = capped;
}

/**
 * Decrement SPD buff remaining turns for a single ally.
 * Reverts +50 SPD when the buff expires.
 */
function decrementSpdBuff(state: SimState, ally: TeamMember): void {
  const key = 'asta_spd_remaining_' + ally.characterId;
  if (!state.stacks[key]) return;
  state.stacks[key]--;
  if (state.stacks[key] <= 0) {
    ally.base_stats[CHAR_SPD_ID] = (ally.base_stats[CHAR_SPD_ID] || 100) - 50;
    state.stacks[key] = 0;
    state.addLog({
      type: 'event',
      message: `Asta: SPD +50 expired for ${ally.name} (SPD now ${(ally.base_stats[CHAR_SPD_ID]).toFixed(0)})`
    });
  }
}

// ── Kit ───────────────────────────────────────────────────────────────────────

export const Asta: CharacterKit = {
  id: ASTA_ID,
  name: "Asta",
  path: "Harmony",
  element: "Fire",
  slot_names: {
    basic:   "Spectrum Beam",
    skill:   "Meteor Storm",
    ultimate:"Astral Blessing",
    talent:  "Astrometry",
  },
  abilities: {
    basic: {
      default_multiplier: 1.0,        // Lv.6: 100% ATK
      stat_id: ATK_ID,
      targetType: 'SingleTarget',
      toughness_damage: 10
    },
    skill: {
      // First hit only; the remaining 4 (or 5 at E1) bounces are fired in onAfterAction
      default_multiplier: 0.5,        // Lv.10: 50% ATK per bounce
      stat_id: ATK_ID,
      targetType: 'Bounce',
      toughness_damage: 10
    },
    ultimate: {
      default_multiplier: 0,          // No direct DMG  Epure support
      stat_id: ATK_ID,
      targetType: 'Support'
    },
    talent: {
      default_multiplier: 0,
      stat_id: ATK_ID
    }
  },

  hooks: {
    // ── Battle Start ───────────────────────────────────────────────────────────
    onBattleStart: (state, member) => {
      state.stacks[CHARGING_KEY] = 0;
      state.stacks[TURN_KEY]     = 0;

      // A4: All Fire-element allies +18% Fire DMG (Asta herself is Fire so she also benefits)
      state.team.forEach(m => {
        if (m.element === 'Fire') {
          m.buffs.dmg_boost += 18;
        }
      });
      state.addLog({ type: 'event', message: `Asta A4: Fire DMG +18% applied to all Fire allies.` });
    },

    // ── Turn Start ─────────────────────────────────────────────────────────────
    onTurnStart: (state, member) => {
      state.stacks[TURN_KEY] = (state.stacks[TURN_KEY] || 0) + 1;

      // From 2nd turn onward: reduce Charging stacks
      if (state.stacks[TURN_KEY] >= 2) {
        if (state.stacks[E2_SKIP_KEY]) {
          // E2: skip reduction this turn
          state.stacks[E2_SKIP_KEY] = 0;
          state.addLog({ type: 'event', message: `Asta E2: Charging reduction skipped this turn.` });
        } else {
          const reduction = member.eidolon >= 6 ? 2 : 3; // E6 reduces by 1 less
          const current = state.stacks[CHARGING_KEY] || 0;
          if (current > 0) {
            const next = Math.max(0, current - reduction);
            setCharging(state, next);
            state.addLog({
              type: 'event',
              message: `Asta Talent: Charging -${reduction} at turn start (${current} ↁE${next}).`
            });
          }
        }
      }

      const stacks = state.stacks[CHARGING_KEY] || 0;
      if (stacks > 0) {
        state.addLog({
          type: 'event',
          message: `Asta Talent: ${stacks} Charging stacks  EAll allies +${stacks * 14}% ATK.`
        });
      }
    },

    // ── Before Action ─────────────────────────────────────────────────────────
    onBeforeAction: (state, member, action) => {
      if (action.type === 'skill') {
        // Simulator will add +30 energy after this action; we need net +6.
        // Pre-subtract 24 so that: (current - 24) + 30 = current + 6.
        state.stacks[ASTA_ID] = (state.stacks[ASTA_ID] || 0) - 24;

        // E4: +15% ERR when 2+ stacks ↁE~+1 extra (6 ÁE0.15 = 0.9 ≁E1)
        if (member.eidolon >= 4 && (state.stacks[CHARGING_KEY] || 0) >= 2) {
          state.stacks[ASTA_ID] += 1;
        }
      } else if (action.type === 'basic') {
        // E4: +15% ERR when 2+ stacks ↁE+3 extra (20 ÁE0.15 = 3)
        if (member.eidolon >= 4 && (state.stacks[CHARGING_KEY] || 0) >= 2) {
          state.stacks[ASTA_ID] = (state.stacks[ASTA_ID] || 0) + 3;
        }
      }
    },

    // ── After Action ──────────────────────────────────────────────────────────
    onAfterAction: (state, member, action, target) => {

      // ── BASIC ATK ──────────────────────────────────────────────────────────
      if (action.type === 'basic' && target) {
        // Talent: +1 Charging for hitting main target, +1 extra if Fire Weakness
        let chargeGain = 1;
        if (target.weaknesses.includes('Fire')) chargeGain++;
        const before = state.stacks[CHARGING_KEY] || 0;
        setCharging(state, before + chargeGain);
        state.addLog({
          type: 'event',
          message: `Asta Talent: +${chargeGain} Charging from Basic (${before} ↁE${state.stacks[CHARGING_KEY]}).`
        });

        // A2: 80% base chance to Burn for 3 turns
        if (Math.random() < 0.80) {
          // DoT value = 50% of basic ATK DMG (recalculated with current member stats)
          const basicResult = calculateHsrDamage({
            character: member,
            lightcone: member.lightcone,
            enemy: target,
            ability_multiplier: 1.0,
            scaling_stat_id: ATK_ID
          });
          const dotDmg = Math.floor(basicResult.expected_dmg * 0.5);
          target.activeDebuffs['asta_burn'] = { duration: 3, value: dotDmg };
          state.addLog({
            type: 'event',
            message: `Asta A2: Burn on ${target.name}  E${dotDmg.toLocaleString()} DoT/turn for 3 turns.`
          });
        }

        // Decrement Asta's own SPD buff after her action
        decrementSpdBuff(state, member);
      }

      // ── SKILL (Bounce) ─────────────────────────────────────────────────────
      if (action.type === 'skill') {
        const aliveEnemies = state.enemies.filter((e): e is SimEnemy => e !== null && e.hp > 0);
        if (aliveEnemies.length === 0) {
          decrementSpdBuff(state, member);
          return;
        }

        // Snapshot fire-weakness info BEFORE enemies potentially die in bounces
        const fireWeakMap = new Map<string, boolean>();
        aliveEnemies.forEach(e => fireWeakMap.set(e.instanceId, e.weaknesses.includes('Fire')));

        // Track unique enemies hit (main target was already hit by simulator)
        const hitIds = new Set<string>();
        if (target) hitIds.add(target.instanceId);

        // E1: +1 extra bounce (4 ↁE5 extra hits after the first)
        const extraBounces = 4 + (member.eidolon >= 1 ? 1 : 0);
        let totalBounceDmg = 0;

        for (let i = 0; i < extraBounces; i++) {
          // Re-query alive enemies each bounce in case one died
          const currentAlive = state.enemies.filter((e): e is SimEnemy => e !== null && e.hp > 0);
          if (currentAlive.length === 0) break;

          const bounceTarget = currentAlive[Math.floor(Math.random() * currentAlive.length)];
          const result = calculateHsrDamage({
            character: member,
            lightcone: member.lightcone,
            enemy: bounceTarget,
            ability_multiplier: 0.5,
            scaling_stat_id: ATK_ID
          });

          bounceTarget.hp = Math.max(0, Math.floor(bounceTarget.hp - result.expected_dmg));
          state.totalDamage  += result.expected_dmg;
          totalBounceDmg     += result.expected_dmg;
          hitIds.add(bounceTarget.instanceId);

          state.addLog({
            type: 'event',
            message: `Bounce ${i + 1}: ${bounceTarget.name} ↁE${result.expected_dmg.toLocaleString()} DMG (HP: ${bounceTarget.hp.toLocaleString()}/${bounceTarget.max_hp.toLocaleString()})`
          });
        }

        if (state.checkEnemies) state.checkEnemies();

        // Talent: Charging stacks from skill (1 per unique enemy hit + 1 if Fire Weakness)
        let chargeGain = 0;
        hitIds.forEach(id => {
          chargeGain++;
          if (fireWeakMap.get(id)) chargeGain++;
        });

        if (chargeGain > 0) {
          const before = state.stacks[CHARGING_KEY] || 0;
          setCharging(state, before + chargeGain);
          state.addLog({
            type: 'event',
            message: `Asta Talent: +${chargeGain} Charging from Skill (${before} ↁE${state.stacks[CHARGING_KEY]}).`
          });
        }

        state.addLog({
          type: 'event',
          message: `Meteor Storm: ${totalBounceDmg.toLocaleString()} total bounce DMG (${extraBounces} extra hits, ${hitIds.size} unique enemies).`
        });

        // Decrement Asta's own SPD buff after her action
        decrementSpdBuff(state, member);
      }
    },

    // ── Ultimate ──────────────────────────────────────────────────────────────
    onUlt: (state, member) => {
      // Manual energy reset: ult costs 120 and grants 5 energy
      state.stacks[ASTA_ID] = 5;

      // E2: skip next Charging reduction on Asta's turn
      if (member.eidolon >= 2) {
        state.stacks[E2_SKIP_KEY] = 1;
        state.addLog({ type: 'event', message: `Asta E2: Next Charging reduction will be skipped.` });
      }

      // Apply SPD +50 to all alive allies for 2 turns
      const aliveTeam = state.team.filter(m => !m.isDowned);
      aliveTeam.forEach(m => {
        m.base_stats[CHAR_SPD_ID] = (m.base_stats[CHAR_SPD_ID] || 100) + 50;
        state.stacks['asta_spd_remaining_' + m.characterId] = 2;
      });

      const names = aliveTeam.map(m => m.name || m.characterId).join(', ');
      state.addLog({
        type: 'event',
        message: `Astral Blessing: All allies [${names}] gain SPD +50 for 2 turns.`
      });
    },

    // ── Enemy Turn: Apply Burn DoT ─────────────────────────────────────────────
    onEnemyTurnStart: (state, member, enemy) => {
      const burn = enemy.activeDebuffs['asta_burn'];
      if (burn?.value) {
        enemy.hp = Math.max(0, enemy.hp - burn.value);
        state.totalDamage += burn.value;
        state.addLog({
          type: 'event',
          message: `Asta A2: Burn DoT ${burn.value.toLocaleString()} on ${enemy.name} (HP: ${enemy.hp.toLocaleString()}/${enemy.max_hp.toLocaleString()}).`
        });
        if (state.checkEnemies) state.checkEnemies();
      }
    },

    // ── Ally Action: Decrement SPD buff for the acting ally ───────────────────
    onAllyAction: (state, member, source, actionType) => {
      decrementSpdBuff(state, source);
    }
  },

  special_modifiers: {
    energy_type: "ENERGY",
    energy_cost: 120,
    stat_boosts: (stats: any) => ({
      dmg_boost: 22.4,  // Minor trace: Fire DMG +22.4% (Asta herself)
      crit_rate:  6.7,  // Minor trace: CRIT Rate +6.7%
      // DEF +22.5%  Enot tracked in combat buffs struct
    }),
    eidolon_level_boosts: (eidolon: number) => ({
      ...(eidolon >= 3 ? { skill: 2, talent: 2 } : {}),
      ...(eidolon >= 5 ? { ultimate: 2, basic: 1 } : {})
    })
  }
};

File: simulation-api/hsr-optimizer-engine/src/characters/asta.rs
use crate::damage;
use crate::effects;
use crate::ids;
use crate::models::{ActionParams, ActionType, SimState, StatusEffect};

const CHARGING_KEY: &str = "asta_charging";
const E2_SKIP_KEY: &str  = "asta_e2_skip";
const TURN_KEY: &str     = "asta_turn";

fn set_charging(state: &mut SimState, new_stacks: f64) {
    let old = state.stacks.get(CHARGING_KEY).copied().unwrap_or(0.0);
    let capped = new_stacks.clamp(0.0, 5.0);
    let diff = (capped - old) * 14.0;
    if diff.abs() > 0.001 {
        for m in state.team.iter_mut() {
            if !m.is_downed { m.buffs.atk_percent += diff; }
        }
    }
    state.stacks.insert(CHARGING_KEY.to_string(), capped);
}

fn decrement_spd_buff(state: &mut SimState, member_idx: usize) {
    let key = format!("asta_spd_remaining_{}", state.team[member_idx].kit_id);
    let remaining = state.stacks.get(&key).copied().unwrap_or(0.0);
    if remaining <= 0.0 { return; }
    if remaining <= 1.0 {
        // Revert SPD
        let spd = state.team[member_idx].base_stats
            .entry(ids::CHAR_SPD_ID.to_string())
            .or_insert(100.0);
        *spd -= 50.0;
        state.stacks.remove(&key);
    } else {
        state.stacks.insert(key, remaining - 1.0);
    }
}

pub fn on_battle_start(state: &mut SimState, idx: usize) {
    state.team[idx].max_energy        = 120.0;
    state.team[idx].buffs.dmg_boost  += 22.4; // minor trace: Fire DMG +22.4%
    state.team[idx].buffs.def_percent += 22.5; // minor trace: DEF +22.5%
    state.team[idx].buffs.crit_rate  +=  6.7; // minor trace: CRIT Rate +6.7%

    state.stacks.insert(CHARGING_KEY.to_string(), 0.0);
    state.stacks.insert(TURN_KEY.to_string(), 0.0);

    // A4: All Fire allies +18% Fire DMG
    for m in state.team.iter_mut() {
        if m.element == "Fire" {
            m.buffs.dmg_boost += 18.0;
        }
    }
}

pub fn on_turn_start(state: &mut SimState, idx: usize) {
    let turn = state.stacks.get(TURN_KEY).copied().unwrap_or(0.0) + 1.0;
    state.stacks.insert(TURN_KEY.to_string(), turn);

    if turn >= 2.0 {
        if state.stacks.get(E2_SKIP_KEY).copied().unwrap_or(0.0) >= 1.0 {
            state.stacks.insert(E2_SKIP_KEY.to_string(), 0.0);
        } else {
            let eidolon = state.team[idx].eidolon;
            let reduction = if eidolon >= 6 { 2.0 } else { 3.0 };
            let current = state.stacks.get(CHARGING_KEY).copied().unwrap_or(0.0);
            if current > 0.0 {
                set_charging(state, (current - reduction).max(0.0));
            }
        }
    }
}

pub fn on_before_action(
    state: &mut SimState,
    idx: usize,
    action: &mut ActionParams,
    _target_idx: Option<usize>,
) {
    // Correct energy: skill should only gain +6 net (simulator will add +30)
    if action.action_type == ActionType::Skill {
        state.team[idx].energy = (state.team[idx].energy - 24.0).max(0.0);
    }

    // A6: +6% DEF per Charging stack
    let charging = state.stacks.get(CHARGING_KEY).copied().unwrap_or(0.0);
    state.team[idx].buffs.def_percent += charging * 6.0;

    // E4: +15% ERR when Charging >= 2
    if state.team[idx].eidolon >= 4 && charging >= 2.0 {
        state.team[idx].buffs.energy_regen_rate += 15.0;
    }
}

pub fn on_after_action(
    state: &mut SimState,
    idx: usize,
    action: &ActionParams,
    target_idx: Option<usize>,
) {
    // A2 Burn: 50% of Basic ATK DMG per tick
    let eff_atk = state.team[idx].base_stats.get(ids::CHAR_ATK_ID).copied().unwrap_or(0.0)
        * (1.0 + state.team[idx].buffs.atk_percent / 100.0);
    let burn_val = eff_atk * 0.5;
    let ehr = state.team[idx].base_stats.get(ids::CHAR_EHR_ID).copied().unwrap_or(0.0)
            + state.team[idx].buffs.effect_hit_rate;

    match action.action_type {
        ActionType::Basic => {
            if let Some(t) = target_idx {
                if let Some(enemy) = state.enemies[t].as_ref() {
                    let fire_weak = enemy.weaknesses.contains(&"Fire".to_string());
                    let gain = if fire_weak { 2.0 } else { 1.0 };
                    let current = state.stacks.get(CHARGING_KEY).copied().unwrap_or(0.0);
                    set_charging(state, current + gain);
                }
                // A2: Burn on Basic hit (80% base chance)
                if let Some(enemy) = state.enemies[t].as_mut() {
                    effects::try_apply_enemy_debuff(ehr, enemy, "asta_burn", StatusEffect {
                        duration: 3,
                        value:    burn_val,
                        stat:     Some("Burn".to_string()),
                        effects:  vec![],
                    }, 0.8);
                }
            }
            decrement_spd_buff(state, idx);
        }
        ActionType::Skill => {
            // Fire 4 extra bounce hits (5 at E1)
            let eidolon = state.team[idx].eidolon;
            let extra_bounces = if eidolon >= 1 { 5 } else { 4 };
            let member = state.team[idx].clone();
            let bounce_action = ActionParams {
                action_type:      ActionType::Skill,
                scaling_stat_id:  ids::CHAR_ATK_ID.to_string(),
                multiplier:       0.5,
                extra_multiplier: 0.0,
                extra_dmg:        0.0,
                toughness_damage: 0.0,
                inflicts_debuff:  false,
                is_ult_dmg:       false,
            };

            let mut hit_ids: Vec<String> = Vec::new();
            if let Some(t) = target_idx {
                if let Some(e) = state.enemies[t].as_ref() {
                    hit_ids.push(e.instance_id.clone());
                }
            }

            for k in 0..extra_bounces {
                // Find alive enemies
                let alive: Vec<usize> = state.enemies.iter().enumerate()
                    .filter_map(|(i, s)| if s.as_ref().map_or(false, |e| e.hp > 0.0) { Some(i) } else { None })
                    .collect();
                if alive.is_empty() { break; }
                // Prefer not-yet-hit enemies
                let pick = alive.iter()
                    .find(|&&i| !hit_ids.contains(
                        &state.enemies[i].as_ref().map(|e| e.instance_id.clone()).unwrap_or_default()
                    ))
                    .copied()
                    .unwrap_or(alive[k % alive.len()]);

                let bounce_dmg = state.enemies[pick].as_ref()
                    .map(|e| {
                        hit_ids.push(e.instance_id.clone());
                        damage::calculate_damage(&member, e, &bounce_action)
                    })
                    .unwrap_or(0.0);
                if bounce_dmg > 0.0 {
                    if let Some(e) = state.enemies[pick].as_mut() { e.hp -= bounce_dmg; }
                    state.total_damage += bounce_dmg;
                }
                if state.enemies[pick].as_ref().map_or(false, |e| e.hp <= 0.0) {
                    state.enemies[pick] = None;
                }
            }

            // Charging stacks: 1 per unique enemy hit + 1 if Fire Weakness
            let fire_key = "Fire".to_string();
            let mut gain = 0.0f64;
            for iid in &hit_ids {
                gain += 1.0;
                // Check fire weakness for this enemy instance
                let is_fire_weak = state.enemies.iter()
                    .filter_map(|s| s.as_ref())
                    .find(|e| &e.instance_id == iid)
                    .map_or(false, |e| e.weaknesses.contains(&fire_key));
                if is_fire_weak { gain += 1.0; }
            }
            let current = state.stacks.get(CHARGING_KEY).copied().unwrap_or(0.0);
            set_charging(state, current + gain);

            decrement_spd_buff(state, idx);
        }
        _ => {}
    }
}

pub fn on_ult(state: &mut SimState, idx: usize) {
    state.team[idx].stacks.insert("_ult_handled".to_string(), 1.0);
    state.team[idx].energy = 5.0;

    // E2: skip next charging reduction
    if state.team[idx].eidolon >= 2 {
        state.stacks.insert(E2_SKIP_KEY.to_string(), 1.0);
    }

    // Apply SPD +50 to all alive allies for 2 turns (2 of their own actions)
    let kit_ids: Vec<String> = state.team.iter()
        .filter(|m| !m.is_downed)
        .map(|m| m.kit_id.clone())
        .collect();

    for kit_id in &kit_ids {
        // Find member index
        if let Some(midx) = state.team.iter().position(|m| &m.kit_id == kit_id) {
            *state.team[midx].base_stats.entry(ids::CHAR_SPD_ID.to_string()).or_insert(100.0) += 50.0;
            let key = format!("asta_spd_remaining_{}", kit_id);
            state.stacks.insert(key, 2.0);
        }
    }
}

pub fn on_global_debuff(
    _state: &mut SimState,
    _idx: usize,
    _source_idx: usize,
    _enemy_idx: usize,
) {}

pub fn on_enemy_turn_start(state: &mut SimState, _idx: usize, enemy_idx: usize) {
    // Apply Burn DoT
    let burn_val = state.enemies[enemy_idx].as_ref()
        .and_then(|e| e.active_debuffs.get("asta_burn"))
        .map(|b| b.value)
        .unwrap_or(0.0);
    if burn_val <= 0.0 { return; }

    if let Some(e) = state.enemies[enemy_idx].as_mut() {
        e.hp -= burn_val;
        if e.hp <= 0.0 { state.enemies[enemy_idx] = None; }
    }
    state.total_damage += burn_val;
}

pub fn on_enemy_action(_state: &mut SimState, _idx: usize, _enemy_idx: usize) {}

pub fn on_ally_action(
    state: &mut SimState,
    _idx: usize,
    source_idx: usize,
    _action: &ActionParams,
    _target_idx: Option<usize>,
) {
    decrement_spd_buff(state, source_idx);
}
